enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,307 @@ package hit_tracker_client
+	invocationID    string
+	requestMetadata *repb.RequestMetadata
+	digest          *repb.Digest
+	start           time.Time
+	client          hitpb.HitTrackerServiceClient
 }
 
 func (t *TransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
+	hit := &hitpb.CacheHit{
+		RequestMetadata: t.requestMetadata,
+		Resource: &rspb.ResourceName{
+			Digest:    t.digest,
+			CacheType: rspb.CacheType_CAS,
+		},
+		SizeBytes: bytesTransferredClient,
+		Duration:  durationpb.New(time.Since(t.start)),
+	}
+	t.enqueueFn(t.ctx, t.requestMetadata, hit)
 	return nil
 }
 
 func (t *TransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
-	return nil
+	return status.InternalError("Unxpected call to hit_tracker_client.Record()")
+}
+
+func (h *HitTrackerClient) TrackDownload(digest *repb.Digest) interfaces.TransferTimer {
+	return &TransferTimer{
+		ctx:             h.ctx,
+		enqueueFn:       h.enqueueFn,
+		requestMetadata: h.requestMetadata,
+		digest:          digest,
+		start:           time.Now(),
+		client:          h.client,
+	}
 }
 
-func (h *HitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+// Writes hit the backing cache, so no need to report on hit-tracking.
+func (h *HitTrackerClient) TrackUpload(digest *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h *HitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+func (h *HitTrackerFactory) runWorker() {
+	for {
+		select {
+		case <-h.quit:
+			h.stopped.Store(false)
+			return
+		case <-h.ticker:
+			h.sendTrackRequests(context.Background())
+		}
+	}
+}
+
+// Sends one round of HitTrackerService.Track requests from the queue. This
+// funciton will send one RPC per group that has any pending hit-updates, and
+// return the number of hit-updates sent.
+func (h *HitTrackerFactory) sendTrackRequests(ctx context.Context) int {
+	i := 0
+	groups := make([]groupID, len(h.hitsByGroup))
+	h.mu.Lock()
+	for group := range h.hitsByGroup {
+		groups[i] = group
+		i++
+	}
+	h.mu.Unlock()
+
+	for _, groupID := range groups {
+		h.mu.Lock()
+		headers := h.hitsByGroup[groupID].authHeaders
+		trackRequest := hitpb.TrackRequest{Hits: h.hitsByGroup[groupID].hits}
+		h.hitsByGroup[groupID] = &cacheHits{}
+		h.mu.Unlock()
+
+		if len(trackRequest.Hits) > 0 {
+			ctx = authutil.AddAuthHeadersToContext(ctx, headers, h.authenticator)
+			_, err := h.client.Track(ctx, &trackRequest)
+			metrics.RemoteAtimeUpdatesSent.With(
+				prometheus.Labels{
+					metrics.GroupID:     string(groupID),
+					metrics.StatusLabel: gstatus.Code(err).String(),
+				}).Inc()
+			if err != nil {
+				log.CtxWarningf(ctx, "Error sending Track request to record cache hit-tracking state group %s: %v", groupID, err)
```

Comment:
retry on error?

Reply:
ehh maybe we can just keep an eye on the warnings for now. To make it safely retryable I think we'd need these to be idempotent, which they probably aren't currently